### PR TITLE
#1095 by program refactor 1

### LINKED ISSUE #1095: ByProgramModal refactor
## Describe

`ByProgramModal` component was quite rushed and is confusing. There is an issue #1039 which requires a small change, but I think it's a good opportunity to rewrite/refactor the component.

### Expected

The component is going to have quite a complex state regardless of implementation. State needs to hold the current data that has been selected, the modal data for at least the current step, but ideally any previous steps (selections like order type are not trivial to compute), as if there is no data to select from for the current step, a user should have feedback. i.e. no periods for the previously selected program+order type. So using a reducer hook should help manage this.

### Version and device info

- App version: 2.3+

### Additional context

Have done some work on this already. To break it up there will be PR's for

- [ ] `Step` component
- [ ] New `byProgramModal`, including a reducer
- [ ] Bringing it together

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #1095, #1039 

## Change summary

Adds a `Step` component to replace `SequentialStep`.

This component is stateful with the `modalData` for the modal which opens on press. This is to use for determining if the step is in an error state or not. On press, the component passes this data back to `ByProgramModal` for opening the modal. Modal data is memoize'd as the data retrieval can be non-trivial.

A part of a few PR's for #1095, this should be the first review.

## Testing
Will be in the final PR - #1098

### Related areas to think about
Will not really do anything for now.

### CODE REVIEW

File: src/widgets/Step.js [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+ * and stored for performance as to not re-fetch every re-render
+ */
+export const Step = memo(props => {
+  const [modalData, setModalData] = useState([]);
+  const { data, field, type, stepKey, status, getModalData, onPress } = props;
+  // Each stepKey used with this component has a placeholder/error string in getLocalisation.
+  const { placeholder, errorString } = getLocalisation({ stepKey });
+  // Calculate the errorState in this component after fetching modalData. If there is no
+  // modalData (there are no options to choose from), display the error string and a new icon.
+  const errorState = !(modalData && modalData.length) && status === 'CURRENT' && getModalData;
```
Comment (wlthomson) [inline src/widgets/Step.js]: Looks like this is a boolean, maybe `isErrorState` or something?

File: src/widgets/Step.js [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    return <View style={iconContainerStyle}>{icon}</View>;
+  };
+
+  // Render a pencil on text editing or a down arrow for list selections.
+  const EditIcon = () => {
+    if (errorState) return null;
+    return (
+      <View>
+        {type === 'input' && <FontAwesome name="pencil" size={20} color="white" />}
+        {type !== 'input' && <Ionicons name="ios-arrow-down" size={20} color="white" />}
```
Comment (wlthomson) [inline src/widgets/Step.js]: Same as before, maybe components for these?

Then could even push it to a one-liner:

```
const EditIcon = () => <View>{errorState && (type === 'input' ? pencilIcon : downArrowIcon)}</View>

File: src/widgets/Step.js [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+      </View>
+    );
+  };
+
+  // Render the error string if in an error state, the title if there is
+  // data/the step is complete and a placeholder otherwise.
+  const TextDisplay = () => {
+    const { textContainerStyle, textStyle } = localStyles;
+    const text = (data && data[field]) || data;
```
Comment (wlthomson) [inline src/widgets/Step.js]: Maybe a target for future refactoring to make data always exclusively either `string` or `object`?

File: src/widgets/Step.js [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+ * @prop {Object} data - the data object the step represents. Used for displaying details
+ * @prop {String} field - The field in the data object to display
+ * @prop {String} type  - The type of step, select (dropdown) or input (text field).
+ * @prop {String} stepKey - A string key for this step.
+ * @prop {String} status - The status of this step (as above)
+ * @prop {Func}   getModalData - Callback for fetching this steps modal data
+ * @prop {Func}   onPress      - Callback for pressing this step (opens modal)
+ * @state modalData - Data to pass to the modal on opening and for determining if the
+ * current step is in an error state. Fetched when the status of this step changes,
+ * and stored for performance as to not re-fetch every re-render
```
Comment (wlthomson) [inline src/widgets/Step.js]: A few minor changes to the docstring, maybe we compromise on the whitespace and go with:

```
/**
 * Component rendering a sequential step in a process, according to it's
 * current status.
 *
 * COMPLETE = The step is complete, render a tick and is pressable.
 * CURRENT = The current step to complete, render an arrow and is pressable.
 * INCOMPLETE = A future step, no icon and no pressable.
 * ERROR = A step not able to be completed, render an X and is not pressable.
 *
 * @prop {Object} data - the data object the step represents. Used for displaying details
 * @prop {String} field - The field in the data object to display
 * @prop {String} type  - The type of step, select (dropdown) or input (text field).
 * @prop {String} stepKey - A string key for this step.
 * @prop {String} status - The status of this step (as above)
 * @prop {Func}   getModalData - Callback for fetching this steps modal data
 * @prop {Func}   onPress      - Callback for pressing this step (opens modal)
 * @state {Array} modalData - Data to pass to the modal on opening and for determining if the
 * current step is in an error state. Fetched when the status of this step changes,
 * and stored for performance as to not re-fetch every re-render
 */
```

File: src/widgets/Step.js [hunk 5]
```diff
@@ -0,0 +1,158 @@
+/* eslint-disable react/forbid-prop-types */
```
Comment (wlthomson) [inline src/widgets/Step.js]: Do we need the `forbid-prop-types`?

File: src/widgets/Step.js [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+  return { placeholder: placeholders[stepKey], errorString: errorString[stepKey] };
+};
+
+/**
+ * Component rendering a sequential step in a process, according to it's
+ * current status.
+ * COMPLETE = The step is complete, render a tick and is pressable.
+ * CURRENT = The current step to complete, render an arrow and is pressable.
+ * INCOMPLETE = A future step, no icon and no pressable.
+ * ERROR = A step not able to be completed, render an X and is not pressable.
```
Comment (wlthomson) [inline src/widgets/Step.js]: Could these be defined in a const enum somewhere?

E.g.

```
const stepStatus = Object.freeze({
  'COMPLETE': 1,
  'CURRENT': 2,
  ...
});
```

Then you can do stuff like:

```
if (status === stepStatus.CURRENT) {
  ...
}
```

Probably a better name, can't think of one off top of my head!


```

[the remark this answers is not in the document]
Comment (joshxg) [reply src/widgets/Step.js]: Is now as per suggestions

File: src/widgets/Step.js [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+  /** Inner components */
+
+  // Render a red X on error state, nothing on incomplete status, a green tick when
+  // complete or a orange arrow on the current step.
+  const StepIcon = () => {
+    const { iconContainerStyle } = localStyles;
+    let icon;
+    if (errorState) icon = <Ionicons name="ios-close" color="red" size={30} />;
+    else if (status === 'INCOMPLETE') icon = null;
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/widgets/Step.js]: nice 6c00798e69462ab46abae4eb0cc8e72686723c4f (screwed this commit up oopsies)

This includes changes for https://github.com/openmsupply/mobile/pull/1096#discussion_r313702938 and https://github.com/openmsupply/mobile/pull/1096#discussion_r313705861

File: src/widgets/Step.js [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+    else if (status === 'COMPLETE') icon = <Ionicons name="md-checkmark" color="green" size={30} />;
+    else if (status === 'CURRENT') {
+      icon = <Ionicons name="md-arrow-round-forward" color={SUSSOL_ORANGE} size={30} />;
+    }
+
+    return <View style={iconContainerStyle}>{icon}</View>;
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/widgets/Step.js]: https://github.com/openmsupply/mobile/pull/1096#discussion_r315009986

File: src/widgets/Step.js (hunk 2 above)
Comment (joshxg) [reply src/widgets/Step.js]: https://github.com/openmsupply/mobile/pull/1096#discussion_r315009986 nice

File: src/widgets/Step.js (hunk 1 above)
Comment (joshxg) [reply src/widgets/Step.js]: Removed in 6c00798

File: src/widgets/Step.js (hunk 3 above)
Comment (joshxg) [reply src/widgets/Step.js]: I suppose so, makes the code ugly but is a bit easier to be a little flexible

File: src/widgets/Step.js (hunk 4 above)
Comment (joshxg) [reply src/widgets/Step.js]: 9c19b0916c80a178e88998822393244a60675596 cheers

File: src/widgets/Step.js (hunk 5 above)
Comment (joshxg) [reply src/widgets/Step.js]: yeah - I still need to get round to putting it in the eslint config (object)

File: src/widgets/Step.js (hunk 6 above)
Comment (joshxg) [reply src/widgets/Step.js]: really like it and went to write it, but it's going to take a few other changes a little out the scope of this PR. Will make this change in the last PR for the issue. Probably making it exportable to use in the `ByProgramModal` component

**Comment by joshxg** [comment]
@wlthomson Added changes from suggestions so dismissed stale approval

**Comment by Chris-Petty** [comment]
@wlthomson pls resolve convos and mergy

File: src/widgets/Step.js (hunk 6 above)
Comment (wlthomson) [reply src/widgets/Step.js]: Sounds good!

**Pull request merged**